internal/providers/vault/certificate.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+const (
+	VaultJwtAudience    = "vault.namespace.systems"
+	vaultRequestTimeout = 10 * time.Second
+)
+
+func Register() {
+	p := &provider{}
+	combined.RegisterSecretsProvider(
+		func(ctx context.Context, srvRef *secrets.ServerRef, cfg *vault.Certificate) ([]byte, error) {
+			ca := cfg.GetCa()
+			if ca == nil {
+				return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing CA configuration")
+			}
+
+			if err := p.Login(ctx, ca, VaultJwtAudience); err != nil {
+				return nil, err
+			}
+
+			commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(srvRef.RelPath, "/", "-"), cfg.BaseDomain)
+			return p.IssueCertificate(ctx, ca, commonName)
+		},
+	)
+}
+
+type tlsKeyPair struct {
+	PrivateKey  string   `json:"private_key"`
+	Certificate string   `json:"certificate"`
+	CaChain     []string `json:"ca_chain"`
+}
+
+type provider struct {
+	vaultClient *vaultclient.Client
+	once        sync.Once
+}
+
+func (p *provider) Login(ctx context.Context, caCfg *vault.CertificateAuthority, audience string) error {
+	var rErr error
+	p.once.Do(func() {
+		if err := tasks.Return0(ctx, tasks.Action("vault.login").Arg("namespace", caCfg.VaultNamespace).Arg("address", caCfg.VaultAddr),
+			func(ctx context.Context) error {
+				var err error
+				p.vaultClient, err = vaultclient.New(
+					vaultclient.WithAddress(caCfg.VaultAddr),
+					vaultclient.WithRequestTimeout(vaultRequestTimeout),
+				)
+				if err != nil {
+					return fnerrors.InvocationError("vault", "failed to create vault client: %w", err)
+				}
+
+				p.vaultClient.SetNamespace(caCfg.VaultNamespace)
+
+				idTokenResp, err := fnapi.IssueIdToken(ctx, audience, 1)
+				if err != nil {
+					return err
+				}
+
+				loginResp, err := p.vaultClient.Auth.JwtLogin(ctx, schema.JwtLoginRequest{Jwt: idTokenResp.IdToken},
+					vaultclient.WithMountPath(caCfg.GetAuthMethod()),
```

Comment:
Why not just caCfg.AuthMethod?

Reply:
We usually prefer the proto accessor methods due to their nil safety.

Reply:
same as before - we prefer to use methods to avoid unnecessary panics.